tensorflow/contrib/ios_examples/camera/tensorflow_utils.mm

```diff
@@ -119,57 +121,108 @@ bool PortableReadFileToProto(const std::string& file_name,
                              std::unique_ptr<tensorflow::Session>* session) {
   tensorflow::SessionOptions options;
-  
+
   tensorflow::Session* session_pointer = nullptr;
-  tensorflow::Status session_status = tensorflow::NewSession(options, &session_pointer);
+  tensorflow::Status session_status =
+      tensorflow::NewSession(options, &session_pointer);
   if (!session_status.ok()) {
     LOG(ERROR) << "Could not create TensorFlow Session: " << session_status;
     return session_status;
   }
   session->reset(session_pointer);
-  LOG(INFO) << "Session created.";
-  
+
   tensorflow::GraphDef tensorflow_graph;
-  LOG(INFO) << "Graph created.";
-  
+
   NSString* model_path = FilePathForResourceName(file_name, file_type);
   if (!model_path) {
     LOG(ERROR) << "Failed to find model proto at" << [file_name UTF8String]
                << [file_type UTF8String];
     return tensorflow::errors::NotFound([file_name UTF8String],
                                         [file_type UTF8String]);
   }
-  const bool read_proto_succeeded = PortableReadFileToProto(
-    [model_path UTF8String], &tensorflow_graph);
+  const bool read_proto_succeeded =
+      PortableReadFileToProto([model_path UTF8String], &tensorflow_graph);
   if (!read_proto_succeeded) {
     LOG(ERROR) << "Failed to load model proto from" << [model_path UTF8String];
     return tensorflow::errors::NotFound([model_path UTF8String]);
   }
-  
-  LOG(INFO) << "Creating session.";
+
   tensorflow::Status create_status = (*session)->Create(tensorflow_graph);
   if (!create_status.ok()) {
     LOG(ERROR) << "Could not create TensorFlow Graph: " << create_status;
     return create_status;
   }
-  
+
+  return tensorflow::Status::OK();
+}
+
+tensorflow::Status LoadMemoryMappedModel(
+    NSString* file_name, NSString* file_type,
+    std::unique_ptr<tensorflow::Session>* session,
+    std::unique_ptr<tensorflow::MemmappedEnv>* memmapped_env) {
+  NSString* network_path = FilePathForResourceName(file_name, file_type);
+  memmapped_env->reset(
+      new tensorflow::MemmappedEnv(tensorflow::Env::Default()));
+  tensorflow::Status mmap_status =
+      (memmapped_env->get())->InitializeFromFile([network_path UTF8String]);
+  if (!mmap_status.ok()) {
+    LOG(ERROR) << "MMap failed with " << mmap_status.error_message();
+    return mmap_status;
+  }
+
+  tensorflow::GraphDef tensorflow_graph;
+  tensorflow::Status load_graph_status = ReadBinaryProto(
+      memmapped_env->get(),
+      tensorflow::MemmappedFileSystem::kMemmappedPackageDefaultGraphDef,
+      &tensorflow_graph);
+  if (!load_graph_status.ok()) {
+    LOG(ERROR) << "MMap load graph failed with "
+               << load_graph_status.error_message();
+    return load_graph_status;
+  }
+
+  tensorflow::SessionOptions options;
+  // Disable optimizations on this graph so that constant folding doesn't
+  // increase the memory footprint by creating new constant copies of the weight
```

Comment:
Does it do this even if the constant is not used in some other expression?  We should probably make it so that single ImmutableConst nodes don't get constant-folded into single constant nodes.